test/functional/invoke.cpp

```diff
@@ -0,0 +1,277 @@
+/*
+  SLB.Functional
+
+  Copyright Michael Park, 2017
+  Copyright Agustin Berge, 2017
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#include <slb/functional.hpp>
+
+#include <type_traits>
+#include <utility>
+
+#define CATCH_CONFIG_MAIN
+#include "../catch.hpp"
+
+#define CHECK_NESTED(...)                                                      \
+  do {                                                                         \
+    INFO(__FILE__ "(" << __LINE__ << "): " #__VA_ARGS__);                      \
+    check_##__VA_ARGS__;                                                       \
+  } while (false)
+
+// [func.invoke], invoke
+
+// template <class F, class... Args>
+//   invoke_result_t<F, Args...> invoke(F&& f, Args&&... args)
+//     noexcept(is_nothrow_invocable_v<F, Args...>);
+
+// Account for P0012: "Make exception specifications be part of the type
+// system".
+static constexpr bool p0012 = !std::is_same<void(), void() noexcept>::value;
+
+struct C {
+  int obj;
+  C(int val) : obj(val) {}
+
+  int fun(int base) noexcept(p0012) { return base + 0; }
+  int cfun(int base) const noexcept(p0012) { return base + 1; }
+  int lfun(int base) & noexcept(p0012) { return base + 2; }
+  int rfun(int base) && noexcept(p0012) { return base + 3; }
+  int clfun(int base) const& noexcept(p0012) { return base + 4; }
+  int crfun(int base) const&& noexcept(p0012) { return base + 5; }
+};
+
+struct D : C {
+  D(int val) : C(val) {}
+};
+
+template <typename T, bool IsNothrow = true>
+struct smart_ptr {
+  T* ptr;
+  smart_ptr(T* ptr) : ptr(ptr) {}
+  T& operator*() const noexcept(IsNothrow) { return *ptr; }
+};
+template <typename T>
+using smart_ptr_throws = smart_ptr<T, false>;
+
+template <typename T, bool IsNothrow = true>
+struct conv_to {
+  T val;
+  conv_to(T val) : val(val) {}
+  operator T() const noexcept(IsNothrow) { return val; }
+};
+template <typename T>
+using conv_to_throws = conv_to<T, false>;
+
+std::true_type const nothrows{};
+std::false_type const throws{};
+std::integral_constant<bool, p0012> const p0012_nothrows{};
+
+template <typename T>
+T const* addressof(T const& ref) {
+  return &ref;
+}
+
+template <typename R, bool IsNothrow, typename F, typename A1>
+void check_invoke_obj(R&& r,
+                      std::integral_constant<bool, IsNothrow>,
+                      F&& f,
+                      A1&& a1) {
+  CHECK(::addressof(slb::invoke(std::forward<F>(f), std::forward<A1>(a1))) ==
+        ::addressof(r));
+  CHECK(std::is_same<decltype(
+                         slb::invoke(std::forward<F>(f), std::forward<A1>(a1))),
+                     R&&>::value);
+  CHECK(noexcept(slb::invoke(std::forward<F>(f), std::forward<A1>(a1))) ==
+        IsNothrow);
+}
+
+template <typename R, bool IsNothrow, typename F, typename... Args>
+void check_invoke(R&& r,
```

Comment:
I think this should be called `check_invoke_fun` to be consistent with the `check_mem_fun` pattern.